grammar.js

```diff
@@ -144,28 +144,58 @@ module.exports = grammar({
     false: $ => 'false',
 
     character_literal: $ => token(seq(
-      "'",
+      '\'',
       repeat1(choice(
         /[^\\'\n]/,
         /\\./,
         /\\\n/
       )),
-      "'"
+      '\''
     )),
 
-    string_literal: $ => token(choice(
-      seq('"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)), '"'),
-      // TODO: support multiline string literals by debugging the following:
-      // seq('"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)), '"', '+', /\n/, '"', repeat(choice(/[^\\"\n]/, /\\(.|\n)/)))
-    )),
-
-    text_block: $ => token(
-      seq(
-        /"""\s*\n/,
-        optional(/([^"]|("[^"])|(""[^"]))+/),
-        '"""'
-      )
+    string_literal: $ => choice($._string_literal, $._multiline_string_literal),
+    _string_literal: $ => seq(
+      '"',
+      repeat(choice(
+        $.string_fragment,
+        $.escape_sequence,
+      )),
+      '"'
     ),
+    _multiline_string_literal: $ => seq(
+      '"""',
+      repeat(choice(
+        alias($._multiline_string_fragment, $.multiline_string_fragment),
+        $._escape_sequence,
+      )),
+      '"""'
+    ),
+    // Workaround to https://github.com/tree-sitter/tree-sitter/issues/1156
+    // We give names to the token() constructs containing a regexp
+    // so as to obtain a node in the CST.
+    //
+    string_fragment: $ =>
+      token.immediate(prec(1, /[^"\\]+/)),
+    _multiline_string_fragment: () =>
+      prec.right(choice(
+        /[^"]+/,
+        seq(/"[^"]*/, repeat(/[^"]+/))
+      )),
+
+    _escape_sequence: $ =>
```

Comment:
would have been good to add in comment a link to the javascript grammar that you took as inspiration
for those changes.

Reply:
Ahh, fair, but I use this with almost every grammar I've made, would be a bit cumbersome :sweat_smile: I like to think of it as a more universal solution for tree-sitter parsers needing this feature/rule